contributors/design-proposals/cri-dockershim-checkpoint.md

```diff
@@ -0,0 +1,91 @@
+
+## Motivation
+### Why do we need checkpoint?
+
+
+With CRI, Kubelet only passes configurations (SandboxConfig, ContainerConfig and ImageSpec) when creating sandbox, container and image, and only use the reference id to manage them after creation. However, information in configuration is not only needed during creation. 
+
+In the case of dockershim with CNI network plugin, CNI plugins needs the same information from PodSandboxConfig at creation and deletion. 
+
+```
+Kubelet ---------------------------------
+     | RunPodSandbox(PodSandboxConfig)
+     | StopPodSandbox(PodSandboxID)
+     V
+Dockershim-------------------------------
+     | SetUpPod
+     | TearDownPod
+     V
+Network Plugin---------------------------
+     | ADD
+     | DEL
+     V
+CNI plugin-------------------------------
+```
+
+
+In addition, checkpoint helps to improve the reliability of dockershim. With checkpoints, critical information for disaster recovery could be preserved. Kubelet make decisions based on the reported pod states from runtime shims. Dockershim currently gathers states from docker engine. However, in the case of disaster recovery, docker engine may lose all container information, including the reference ids. Without the reference id, kubelet or dockershim could not reclaim allocated resource. For instances, pod IP and port forwarding iptables rules would leak. 
+
+### Why checkpoint in dockershim?
+- CNI specification does not require CNI plugins to be stateful. And CNI specification does not provide interface to retrieve states from CNI plugins. 
+- Currently there is no uniform checkpoint requirements across existing runtime shims.
+- Need to preserve backward compatibility for kubelet.
+- No need for runtime shims to reconcile configuration drift.
+
+## PodSandbox Checkpoint 
+Checkpoint file will be created for each PodSandbox. Files will be placed under `/var/lib/dockershim/sandbox/`. File name will be `PodSandboxID`. File content will be json encoded. Data structure is as follows:
+
+```go
+type Protocol string
+
+// PortMapping is the port mapping configurations of a sandbox.
+type PortMapping struct {
+	// Protocol of the port mapping.
+	Protocol *Protocol `json:"protocol,omitempty"`
+	// Port number within the container.
+	ContainerPort *int32 `json:"container_port,omitempty"`
+	// Port number on the host.
+	HostPort *int32 `json:"host_port,omitempty"`
+}
+
+// PodSandboxCheckpoint is the checkpoint structure for a sandbox
+type PodSandboxCheckpoint struct{
+	// Pod name of the sandbox. Same as the pod name in the PodSpec.
+	Name *string `json:"name,omitempty"`
+	// Pod namespace of the sandbox. Same as the pod namespace in the PodSpec.
+	Namespace *string `json:"namespace,omitempty"`
+	// Port mappings for the sandbox.
+	PortMappings []*PortMapping `"json: port_mappings,omitempty"`
```

Comment:
Since port mappings are really a facet of the CNI driver, rather than the shim itself, can we describe this struct in a more extensible way?  Like maybe subsystems can extend it in arbitrary ways, and can then extract their own information.

e.g. 

https://play.golang.org/p/cRUgdfehux